lib/web_ui/lib/src/engine/semantics/text_field.dart

```diff
@@ -341,75 +375,103 @@ class TextField extends RoleManager {
+  void _applyTextFromEditingState() {
+    final EditingState editingState = EditingState(
+      text: semanticsObject.value,
+      baseOffset: semanticsObject.textSelectionBase,
+      extentOffset: semanticsObject.textSelectionExtent,
+    );
+    editingState.applyTextToDomElement(editableElement);
+  }
 
-    // Whether we should request that the browser shift focus to the editable
-    // element, so that both the framework and the browser agree on what's
-    // currently focused.
-    bool needsDomFocusRequest = false;
+  void _setEditingState() {
     final EditingState editingState = EditingState(
       text: semanticsObject.value,
       baseOffset: semanticsObject.textSelectionBase,
       extentOffset: semanticsObject.textSelectionExtent,
     );
-    if (semanticsObject.hasFocus) {
-      if (!_hasFocused) {
-        _hasFocused = true;
-        SemanticsTextEditingStrategy.instance.activate(this);
-        needsDomFocusRequest = true;
-      }
-      if (domDocument.activeElement != editableElement) {
-        needsDomFocusRequest = true;
-      }
-      // Focused elements should have full text editing state applied.
-      SemanticsTextEditingStrategy.instance.setEditingState(editingState);
-    } else if (_hasFocused) {
+    SemanticsTextEditingStrategy.instance.setEditingState(editingState);
+  }
+
+  void _invokeIosWorkaround() {
+    if (editableElement != null) {
+      return;
+    }
+
+    _createEditableElement();
+    activeEditableElement.style.transform = 'translate(-999px,-999px)';
+    _positionInputElementTimer?.cancel();
+    _positionInputElementTimer = Timer(_delayBeforePlacement, () {
+      editableElement?.style.transform = '';
+    });
+
+    activeEditableElement.focus();
+    semanticsObject.element.removeAttribute('role');
+
+    activeEditableElement.addEventListener('blur',
+        allowInterop((DomEvent event) {
+      _applyTextFromEditingState();
+      semanticsObject.element.setAttribute('role', 'textbox');
+      activeEditableElement.remove();
       SemanticsTextEditingStrategy.instance.deactivate(this);
 
-      // Only apply text, because this node is not focused.
-      editingState.applyTextToDomElement(editableElement);
+      // Focus on semantics element before removing the editable element, so that
+      // the user can continue navigating the page with the assistive technology.
+      semanticsObject.element.focus();
+      editableElement = null;
+    }));
+  }
 
-      if (_hasFocused && domDocument.activeElement == editableElement) {
-        // Unlike `editableElement.focus()` we don't need to schedule `blur`
-        // post-update because `document.activeElement` implies that the
-        // element is already attached to the DOM. If it's not, it can't
-        // possibly be focused and therefore there's no need to blur.
-        editableElement.blur();
-      }
-      _hasFocused = false;
+  @override
+  void update() {
+    if (editableElement != null) {
```

Comment:
I think all the `editableElement != null` checks can be combined into one top-level check, so that we don't need to check it again multiple times.

Reply:
Yep, you're right.